<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Yusha Liu" />


<title>Dissecting tumor transcriptional heterogeneity from multi-tumor single-cell RNA-seq data using GBCD</title>

<script src="site_libs/header-attrs-2.21/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">gbcd workflow</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephenslab/gbcd-workflow">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Dissecting tumor transcriptional
heterogeneity from multi-tumor single-cell RNA-seq data using GBCD</h1>
<h4 class="author">Yusha Liu</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2023-08-08
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>gbcd-workflow/analysis/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.0.5). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomstephenslabgbcdworkflowtreef2b1d24787754c18ab34093d806ddbcdfb7ac452targetblankf2b1d24a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/stephenslab/gbcd-workflow/tree/f2b1d24787754c18ab34093d806ddbcdfb7ac452" target="_blank">f2b1d24</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomstephenslabgbcdworkflowtreef2b1d24787754c18ab34093d806ddbcdfb7ac452targetblankf2b1d24a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/stephenslab/gbcd-workflow/tree/f2b1d24787754c18ab34093d806ddbcdfb7ac452" target="_blank">f2b1d24</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
working directory clean
</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/gbcd_hnscc_intro.Rmd</code>) and
HTML (<code>docs/gbcd_hnscc_intro.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/gbcd-workflow/blob/f2b1d24787754c18ab34093d806ddbcdfb7ac452/analysis/gbcd_hnscc_intro.Rmd" target="_blank">f2b1d24</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2023-08-08
</td>
<td>
workflowr::wflow_publish("gbcd_hnscc_intro.Rmd", verbose = TRUE)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/gbcd-workflow/d40f7d0624738c58682fd5be5d35b43b0e7af974/docs/gbcd_hnscc_intro.html" target="_blank">d40f7d0</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2023-08-08
</td>
<td>
Built the gbcd_hnscc_intro workflowr page.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/gbcd-workflow/blob/f9fe46de16323cdf0a389003dfb16572c50f75f2/analysis/gbcd_hnscc_intro.Rmd" target="_blank">f9fe46d</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2023-08-08
</td>
<td>
workflowr::wflow_publish("gbcd_hnscc_intro.Rmd", verbose = TRUE)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>This vignette shows how to apply “generalized binary covariance
decomposition” (GBCD) to jointly analyze single-cell RNA-seq (scRNA-seq)
data from <em>malignant cells</em> collected from multiple patients
and/or studies, using a head and neck squamous cell carcinoma (HNSCC)
dataset from <a href="https://doi.org/10.1016/j.cell.2017.10.044">Puram
et al. 2017</a>.</p>
<p>GBCD can effectively dissect tumor transcriptional heterogeneity into
patient/study-specific and shared gene expression programs (GEPs). GBCD
is “unsupervised” in that, unlike tumor-by-tumor (e.g., <a
href="https://doi.org/10.1016/j.cell.2017.10.044">Puram et
al. (2017)</a>) and many harmonization approaches (e.g., <a
href="https://portals.broadinstitute.org/harmony">Harmony</a>, <a
href="https://github.com/welch-lab/liger">Liger</a>), it does not use
information about which cell comes from which tumor or study. Instead,
GBCD only requires the combined scRNA-seq data from all tumors, which
are stored as an <span class="math inline">\(N \times J\)</span> matrix
<span class="math inline">\(Y\)</span> of expression values with entries
<span class="math inline">\(y_{ij}\)</span>, where <span
class="math inline">\(i=1,\dots,N\)</span> indexes malignant cells and
<span class="math inline">\(j=1,\dots,J\)</span> indexes genes. In
typical applications, <span class="math inline">\(Y\)</span> contains
log-transformed pseudo-count-modified UMI counts (“log-pc counts”).</p>
<p>GBCD ultimately yields a decomposition of the expression data matrix
<span class="math inline">\(Y\)</span> into matrices <span
class="math inline">\(L\)</span> and <span
class="math inline">\(F\)</span> such that <span class="math inline">\(Y
\approx L F^T\)</span>, or equivalently, <span
class="math display">\[y_{ij} \approx \sum_{k=1}^K l_{ik}
f_{jk}.\]</span> The <span class="math inline">\(K\)</span> components
should be interpretable as GEPs, with <span
class="math inline">\(l_{ik}\)</span> representing the membership of
cell <span class="math inline">\(i\)</span> in GEP <span
class="math inline">\(k\)</span>, and <span
class="math inline">\(f_{jk}\)</span> representing the effect of GEP
<span class="math inline">\(k\)</span> on the expression of gene <span
class="math inline">\(j\)</span>. When <span
class="math inline">\(y_{ij}\)</span> are log-pc counts, each <span
class="math inline">\(f_{jk}\)</span> approximately represents the
log-fold change (LFC) associated with membership in GEP <span
class="math inline">\(k\)</span>, so we refer to the <span
class="math inline">\(f_{jk}\)</span> values as LFCs, and to the vector
of LFCs <span class="math inline">\((f_{1k}, \dots, f_{Jk})^T\)</span>
as the “signature” of GEP <span class="math inline">\(k\)</span>.</p>
<p>We begin our analysis by loading the R packages used in the analysis,
as well as some custom functions implementing the GBCD model fitting and
interpretation.. Then we show how to apply GBCD to the HNSCC dataset
step-by-step.</p>
<pre class="r"><code>library(ggplot2)
library(cowplot)
library(RColorBrewer)
library(ggrepel)
library(pheatmap)
library(gridExtra)
library(Seurat)
library(Matrix)
library(ebnm)
library(flashier)
library(magrittr)
library(ashr)
source(&quot;../code/fit_cov_ebnmf.R&quot;)</code></pre>
</div>
<div id="the-hnscc-dataset" class="section level2">
<h2>The HNSCC dataset</h2>
<p>This data set contains gene expression data for n = 2,176 malignant
cells collected in primary tumors from 10 HNSCC patients, as well as
matching lymph node (LN) metastases from 5 of these patients. <a
href="https://doi.org/10.1016/j.cell.2017.10.044">Puram et
al. (2017)</a> found that each of these 10 patients mapped to a
molecular subtype of HNSCC, whose signatures were previously defined by
analysis of bulk expression data of 279 <a
href="https://doi.org/10.1038%2Fnature14129">TCGA</a> HNSCC tumors.</p>
<p>Unlike more recently generated data sets, the HNSCC data are not UMI
counts; rather, they are read counts produced by SMART-Seq2. Following
<a href="https://doi.org/10.1016/j.cell.2017.10.044">Puram et
al. (2017)</a>, we define the transformed counts as <span
class="math inline">\(y_{ij} = \log_2(1 +
\mathrm{TPM}_{ij}/10)\)</span>, where <span
class="math inline">\(\mathrm{TPM}_{ij}\)</span> is the
transcript-per-million (TPM) value for gene <span
class="math inline">\(j\)</span> in cell <span
class="math inline">\(i\)</span>.</p>
<pre class="r"><code>load(&quot;../hnscc/hnscc.RData&quot;)
dim(Y)
print(head(info),row.names = FALSE)
# [1]  2176 17113
#                    cell.id lymph.node cancer.cell cell.type sample.id   subject
#     HN26_P14_D11_S239_comb          1           1         0       T26 MEEI26_LN
#     HN26_P25_H09_S189_comb          1           1         0       T26 MEEI26_LN
#     HN26_P14_H06_S282_comb          1           1         0       T26 MEEI26_LN
#     HN25_P25_C04_S316_comb          1           1         0       T25 MEEI25_LN
#     HN26_P25_C09_S129_comb          1           1         0       T26 MEEI26_LN
#  HNSCC26_P24_H05_S377_comb          1           1         0       T26 MEEI26_LN
#   subtype
#  Atypical
#  Atypical
#  Atypical
#     Basal
#  Atypical
#  Atypical</code></pre>
</div>
<div id="visualize-hnscc-data-using-t-sne" class="section level2">
<h2>Visualize HNSCC data using <em>t</em>-SNE</h2>
<p>We plot the 2-D <em>t</em>-SNE embedding of these <span
class="math inline">\(2,176\)</span> malignant cells, which are colored
by patient of origin and tumor stage (primary tumor, LN metastasis), and
shaped by tumor molecular subtype. We set seed to make the
<em>t</em>-SNE reproducible.</p>
<pre class="r"><code>set.seed(100)
hnscc &lt;- CreateSeuratObject(counts = t(Y), project = &quot;HNSCC&quot;, meta.data = info)
hnscc &lt;- FindVariableFeatures(hnscc, selection.method = &quot;vst&quot;,
                              nfeatures = 5000)
all.genes &lt;- rownames(hnscc)
hnscc &lt;- ScaleData(hnscc, features = all.genes)
hnscc &lt;- RunPCA(hnscc, features = VariableFeatures(object = hnscc),
                npcs = 50, verbose = FALSE)
hnscc &lt;- RunTSNE(hnscc, dims = 1:50)</code></pre>
<p>Due to the presence of strong inter-tumor heterogeneity, these cells
demonstrate strong patient effects that are typical of cancer data; the
major structure in the <em>t</em>-SNE visualization is the clustering of
the cells by patient.</p>
<pre class="r"><code>DimPlot(hnscc, label = TRUE, repel = TRUE, pt.size = 1, reduction = &quot;tsne&quot;,
        group.by = &quot;subject&quot;, shape.by = &quot;subtype&quot;, cols = subject_col) +
  guides(shape = guide_legend(override.aes = list(size = 3)), ncol = 1) +
  theme(text = element_text(size = 10)) +
  scale_shape_manual(values = c(15,16,18))</code></pre>
<p><img src="figure/gbcd_hnscc_intro.Rmd/plot-tSNE-1.png" width="600" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plot-tSNE-1">
Past versions of plot-tSNE-1.png
</button>
</p>
<div id="fig-plot-tSNE-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/gbcd-workflow/blob/d40f7d0624738c58682fd5be5d35b43b0e7af974/docs/figure/gbcd_hnscc_intro.Rmd/plot-tSNE-1.png" target="_blank">d40f7d0</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2023-08-08
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="apply-gbcd-to-estimate-gep-memberships-and-signatures"
class="section level2">
<h2>Apply GBCD to estimate GEP memberships and signatures</h2>
<p>This is accomplished with a call to the custom function
<code>flash_fit_cov_ebmf</code>.</p>
<p>To fit GBCD, you need to specify <code>Kmax</code>, an upper bound of
the number of components <span class="math inline">\(K\)</span>, for
initialization; the number of GEPs returned by GBCD is close to but
often not equal to (and can be larger than) <code>Kmax</code>. Generally
speaking, a larger <code>Kmax</code> allows for identifying finer
structure in tumor transcriptional heterogeneity at the expense of
higher computational cost. A good choice of <code>Kmax</code> depends on
the complexity of the analyzed data (e.g., the number of patients and
studies contained), and is often unknown in advance. You1 are encouraged
to explore and compare the results for different values of
<code>Kmax</code>. After some initial exploration, we set
<code>Kmax</code> to 24.</p>
<p>You also need to specify a prior for GEP membership values. GBCD
assigns a “generalized binary” (GB) prior independently to each entry of
<span class="math inline">\(L\)</span>, <span
class="math display">\[l_{ik}\sim (1-\pi_k^l) \delta_0 + \pi_k^l \:
N_{+} (\mu_k,
\sigma_k^2),\]</span> where the ratio <span
class="math inline">\(\sigma_k/\mu_k\)</span> is fixed at some
pre-specified small value. In this analysis, we set <span
class="math inline">\(\sigma_k/\mu_k = 0.04\)</span> for all <span
class="math inline">\(k\)</span>.</p>
<pre class="r"><code>fit.gbcd &lt;-
  flash_fit_cov_ebnmf(Y = Y, Kmax = 24,
                      prior = flash_ebnm(prior_family = &quot;generalized_binary&quot;,
                                         scale = 0.04),
                      extrapolate = FALSE)</code></pre>
<p>Note that you can choose other nonnegative priors for GEP membership
values so long as they are defined in the <a
href="https://stephenslab.github.io/ebnm/">ebnm</a> package. For
example, you can chose a point-exponential prior by setting
<code>prior = ebnm::ebnm_point_exponential</code>.</p>
<p>The time to fit GBCD depends on the choice of <code>Kmax</code> and
the size of the data set being analyzed. It took us a few hours to fit
the GBCD model with <code>Kmax = 24</code> to the HNSCC data set.</p>
<p>Since it takes a while to fit GBCD to the HNSCC dataset, we directly
load in the saved output from the above call and show how to interpret
the results.</p>
<pre class="r"><code>load(&quot;../hnscc/hnscc_gbcd.RData&quot;)</code></pre>
<p>The <code>flash_fit_cov_ebnmf</code> output contains the membership
estimates, stored as an <span class="math inline">\(n \times K\)</span>
matrix,</p>
<pre class="r"><code>dim(fit.gbcd$L)
fit.gbcd$L[1:4,1:5]
# [1] 2176   29
#                         Baseline          GEP1          GEP2          GEP3
# HN26_P14_D11_S239_comb 0.7404634 3.538229e-110 2.323954e-141  9.039494e-01
# HN26_P25_H09_S189_comb 0.6666176 5.623800e-137 9.582078e-135  7.786085e-01
# HN26_P14_H06_S282_comb 0.7976365 8.636394e-136  2.833297e-01  7.258564e-01
# HN25_P25_C04_S316_comb 0.4193556 6.158516e-137  5.124139e-01 4.286821e-135
#                                 GEP4
# HN26_P14_D11_S239_comb 4.589753e-141
# HN26_P25_H09_S189_comb 9.840882e-136
# HN26_P14_H06_S282_comb 2.828108e-122
# HN25_P25_C04_S316_comb 1.527244e-135</code></pre>
<p>and posterior statistics for the LFC estimates, including the
posterior means, stored as an <span class="math inline">\(m \times
K\)</span> matrix,</p>
<pre class="r"><code>dim(fit.gbcd$F$lfc)
fit.gbcd$F$lfc[1:4,1:5]
# [1] 17113    29
#              Baseline          GEP1         GEP2          GEP3          GEP4
# C9orf152 0.0001609319  0.0002929911  0.000234652 -0.0004093085  0.0002814196
# ELMO2    2.0116827172 -0.0060056851 -0.601995065  0.0387999604 -0.0188075633
# PNMA1    1.6386413103 -0.0021806303 -1.244428917  0.0164347046 -0.0059650037
# MMP2     0.4807052706 -0.3115376229  1.673189435 -0.4256140576 -1.1229331251</code></pre>
</div>
<div id="gep-membership-estimates" class="section level2">
<h2>GEP membership estimates</h2>
<p>We now examine at the entries of the <code>L</code> matrix giving
membership estimates of cell <span class="math inline">\(i\)</span> in
GEP <span class="math inline">\(k\)</span>. Here we visually compare the
memberships to the patient labels and provided tumor subtype annotations
with a heatmap.</p>
<pre class="r"><code>anno &lt;- data.frame(subject = info$subject,
                   subtype = info$subtype)
rownames(anno) &lt;- rownames(fit.gbcd$L)
anno_colors &lt;- list(subject = subject_col,
                    subtype = subtype_col)
cols &lt;- colorRampPalette(c(&quot;gray96&quot;,&quot;red&quot;))(50)
brks &lt;- seq(0,1,0.02)
rows &lt;- order(anno$subject)
p &lt;- pheatmap(fit.gbcd$L[rows, -1], cluster_rows = FALSE, cluster_cols = FALSE,
         show_rownames = FALSE, annotation_row = anno,
         annotation_colors = anno_colors, annotation_names_row = FALSE,
         angle_col = 45, fontsize = 8, color = cols, breaks = brks,
         main = &quot;&quot;)
print(p)         </code></pre>
<p><img src="figure/gbcd_hnscc_intro.Rmd/plot-L-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plot-L-1">
Past versions of plot-L-1.png
</button>
</p>
<div id="fig-plot-L-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/gbcd-workflow/blob/d40f7d0624738c58682fd5be5d35b43b0e7af974/docs/figure/gbcd_hnscc_intro.Rmd/plot-L-1.png" target="_blank">d40f7d0</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2023-08-08
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This heatmap shows the membership values of the n = 2,176 cells (the
rows in the heatmap) and the 28 GEPs (the columns in the heatmap). (The
membership values for the “baseline” GEP are not shown.) Cells are
arranged top-to-bottom by tumor molecular subtype and patient of origin.
We rescaled the membership values separately for each GEP so that the
maximum membership for each GEP is always 1.</p>
<p>From this heatmap, we see that GEPs 1, 2 and 3 correspond closely to
the molecular subtypes previously defined based on bulk RNA-seq data:
GEP1 is largely active only in cells from the 2 classical patients, GEP2
is mainly active in cells from the 7 basal patients, and GEP3 is mainly
active in cells from the 1 atypical patient. This demonstrates that GBCD
can reconstruct the molecular subtype information from the single cell
data alone.</p>
<p>Among other GEPs, some are active predominantly in an individual
patient and can thus be interpreted as patient-specific GEPs. We
interpret GEPs 4, 7, 9, 11, 25, 27 as “patient-specific GEPs”; the
remaining GEPs are shared across multiple patients but are capturing
something different than these previously defined molecular
subtypes.</p>
</div>
<div id="gep-signature-estimates" class="section level2">
<h2>GEP signature estimates</h2>
<p>The GEP signature estimates <span
class="math inline">\(f_{jk}\)</span> represent approximately the
log-fold change (using t he base-2 logarithm) associated with membership
in GEP <span class="math inline">\(k\)</span>. We take the approach
commonly used in differential expression analysis and create a “volcano
plot” to visualize the gene signature for a GEP. As an example, this is
the volcano plot for GEP2:</p>
<pre class="r"><code>k &lt;- &quot;GEP2&quot;
pdat &lt;- data.frame(gene = rownames(fit.gbcd$F$lfc), 
                   lfc  = fit.gbcd$F$lfc[,k], 
                   z    = abs(fit.gbcd$F$z_score[,k]), 
                   lfsr = fit.gbcd$F$lfsr[,k],
                   stringsAsFactors = FALSE)
pdat &lt;- transform(pdat,lfsr = cut(lfsr, c(-1,0.001,0.01,0.05,Inf)))
rows  &lt;- with(pdat, which(!(abs(lfc) &gt; quantile(abs(lfc),0.999) | (z &gt; 10))))
pdat[rows, &quot;gene&quot;] &lt;- &quot;&quot;
ggplot(pdat, aes(x = lfc, y = z, color = lfsr, label = gene)) +
  geom_point() + 
  geom_text_repel(color = &quot;black&quot;, size = 2.3, segment.color = &quot;black&quot;,
                  segment.size = 0.25, min.segment.length = 0,
                  max.overlaps = Inf, na.rm = TRUE) +
  scale_color_manual(values = c(&quot;coral&quot;,&quot;orange&quot;,&quot;gold&quot;,&quot;deepskyblue&quot;)) +
  labs(x = &quot;log-fold change&quot;, y = &quot;|posterior z-score|&quot;) + 
  theme_cowplot(font_size = 10)</code></pre>
<p><img src="figure/gbcd_hnscc_intro.Rmd/plot-lfc-1.png" width="600" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plot-lfc-1">
Past versions of plot-lfc-1.png
</button>
</p>
<div id="fig-plot-lfc-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/gbcd-workflow/blob/d40f7d0624738c58682fd5be5d35b43b0e7af974/docs/figure/gbcd_hnscc_intro.Rmd/plot-lfc-1.png" target="_blank">d40f7d0</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2023-08-08
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The significance measure here is not a <em>p</em>-value, but rather a
<a href="https://doi.org/10.1093/biostatistics/kxw041">local false sign
rate (lfsr)</a>.</p>
</div>
<div id="session-info" class="section level2">
<h2>Session info</h2>
<p>This is the version of R and the packages that were used to generate
these results.</p>
<pre class="r"><code>sessionInfo()
# R version 3.6.2 (2019-12-12)
# Platform: x86_64-apple-darwin15.6.0 (64-bit)
# Running under: macOS Catalina 10.15.7
# 
# Matrix products: default
# BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
# LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
# 
# locale:
# [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# attached base packages:
# [1] stats     graphics  grDevices utils     datasets  methods   base     
# 
# other attached packages:
#  [1] ashr_2.2-57        flashier_0.2.51    magrittr_2.0.1     ebnm_1.0-54       
#  [5] Matrix_1.3-4       Seurat_3.2.3       gridExtra_2.3      pheatmap_1.0.12   
#  [9] ggrepel_0.9.1      RColorBrewer_1.1-2 cowplot_1.1.1      ggplot2_3.3.6     
# 
# loaded via a namespace (and not attached):
#   [1] Rtsne_0.15           colorspace_1.4-1     deldir_1.0-6        
#   [4] ellipsis_0.3.2       ggridges_0.5.2       rprojroot_2.0.3     
#   [7] fs_1.5.2             spatstat.data_3.0-1  farver_2.0.1        
#  [10] leiden_0.3.3         listenv_0.8.0        fansi_0.4.0         
#  [13] codetools_0.2-16     splines_3.6.2        knitr_1.37          
#  [16] polyclip_1.10-0      jsonlite_1.7.2       workflowr_1.7.0.5   
#  [19] ica_1.0-2            cluster_2.1.0        png_0.1-7           
#  [22] uwot_0.1.10          shiny_1.7.4          sctransform_0.3.2   
#  [25] compiler_3.6.2       httr_1.4.2           assertthat_0.2.1    
#  [28] fastmap_1.1.0        lazyeval_0.2.2       cli_3.5.0           
#  [31] later_1.0.0          htmltools_0.5.4      tools_3.6.2         
#  [34] rsvd_1.0.2           igraph_1.2.5         gtable_0.3.0        
#  [37] glue_1.4.2           RANN_2.6.1           reshape2_1.4.3      
#  [40] dplyr_1.0.7          Rcpp_1.0.8           spatstat_1.64-1     
#  [43] softImpute_1.4       scattermore_0.7      jquerylib_0.1.4     
#  [46] vctrs_0.3.8          nlme_3.1-142         lmtest_0.9-38       
#  [49] xfun_0.39.1          stringr_1.4.0        globals_0.13.0      
#  [52] trust_0.1-8          mime_0.8             miniUI_0.1.1.1      
#  [55] lifecycle_1.0.3      irlba_2.3.3          goftest_1.2-2       
#  [58] future_1.18.0        MASS_7.3-51.4        zoo_1.8-7           
#  [61] scales_1.1.0         promises_1.1.0       spatstat.utils_3.0-2
#  [64] parallel_3.6.2       yaml_2.2.0           reticulate_1.16     
#  [67] pbapply_1.5-1        sass_0.4.0           rpart_4.1-15        
#  [70] SQUAREM_2017.10-1    stringi_1.4.3        highr_0.8           
#  [73] deconvolveR_1.2-1    horseshoe_0.2.0      truncnorm_1.0-8     
#  [76] rlang_1.0.6          pkgconfig_2.0.3      matrixStats_0.63.0  
#  [79] invgamma_1.1         evaluate_0.14        lattice_0.20-38     
#  [82] ROCR_1.0-11          purrr_0.3.4          tensor_1.5          
#  [85] labeling_0.3         patchwork_1.0.1      htmlwidgets_1.6.1   
#  [88] tidyselect_1.1.1     RcppAnnoy_0.0.18     plyr_1.8.5          
#  [91] R6_2.4.1             generics_0.0.2       DBI_1.1.0           
#  [94] mgcv_1.8-31          pillar_1.6.2         whisker_0.4         
#  [97] withr_2.5.0          fitdistrplus_1.1-1   mixsqp_0.3-48       
# [100] survival_3.1-8       abind_1.4-5          tibble_3.1.3        
# [103] future.apply_1.6.0   crayon_1.4.1         KernSmooth_2.23-16  
# [106] utf8_1.1.4           plotly_4.10.1        rmarkdown_2.21      
# [109] grid_3.6.2           data.table_1.14.2    git2r_0.29.0        
# [112] digest_0.6.23        xtable_1.8-4         tidyr_1.1.3         
# [115] httpuv_1.5.2         munsell_0.5.0        viridisLite_0.3.0   
# [118] bslib_0.3.1</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()
# R version 3.6.2 (2019-12-12)
# Platform: x86_64-apple-darwin15.6.0 (64-bit)
# Running under: macOS Catalina 10.15.7
# 
# Matrix products: default
# BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
# LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
# 
# locale:
# [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# attached base packages:
# [1] stats     graphics  grDevices utils     datasets  methods   base     
# 
# other attached packages:
#  [1] ashr_2.2-57        flashier_0.2.51    magrittr_2.0.1     ebnm_1.0-54       
#  [5] Matrix_1.3-4       Seurat_3.2.3       gridExtra_2.3      pheatmap_1.0.12   
#  [9] ggrepel_0.9.1      RColorBrewer_1.1-2 cowplot_1.1.1      ggplot2_3.3.6     
# 
# loaded via a namespace (and not attached):
#   [1] Rtsne_0.15           colorspace_1.4-1     deldir_1.0-6        
#   [4] ellipsis_0.3.2       ggridges_0.5.2       rprojroot_2.0.3     
#   [7] fs_1.5.2             spatstat.data_3.0-1  farver_2.0.1        
#  [10] leiden_0.3.3         listenv_0.8.0        fansi_0.4.0         
#  [13] codetools_0.2-16     splines_3.6.2        knitr_1.37          
#  [16] polyclip_1.10-0      jsonlite_1.7.2       workflowr_1.7.0.5   
#  [19] ica_1.0-2            cluster_2.1.0        png_0.1-7           
#  [22] uwot_0.1.10          shiny_1.7.4          sctransform_0.3.2   
#  [25] compiler_3.6.2       httr_1.4.2           assertthat_0.2.1    
#  [28] fastmap_1.1.0        lazyeval_0.2.2       cli_3.5.0           
#  [31] later_1.0.0          htmltools_0.5.4      tools_3.6.2         
#  [34] rsvd_1.0.2           igraph_1.2.5         gtable_0.3.0        
#  [37] glue_1.4.2           RANN_2.6.1           reshape2_1.4.3      
#  [40] dplyr_1.0.7          Rcpp_1.0.8           spatstat_1.64-1     
#  [43] softImpute_1.4       scattermore_0.7      jquerylib_0.1.4     
#  [46] vctrs_0.3.8          nlme_3.1-142         lmtest_0.9-38       
#  [49] xfun_0.39.1          stringr_1.4.0        globals_0.13.0      
#  [52] trust_0.1-8          mime_0.8             miniUI_0.1.1.1      
#  [55] lifecycle_1.0.3      irlba_2.3.3          goftest_1.2-2       
#  [58] future_1.18.0        MASS_7.3-51.4        zoo_1.8-7           
#  [61] scales_1.1.0         promises_1.1.0       spatstat.utils_3.0-2
#  [64] parallel_3.6.2       yaml_2.2.0           reticulate_1.16     
#  [67] pbapply_1.5-1        sass_0.4.0           rpart_4.1-15        
#  [70] SQUAREM_2017.10-1    stringi_1.4.3        highr_0.8           
#  [73] deconvolveR_1.2-1    horseshoe_0.2.0      truncnorm_1.0-8     
#  [76] rlang_1.0.6          pkgconfig_2.0.3      matrixStats_0.63.0  
#  [79] invgamma_1.1         evaluate_0.14        lattice_0.20-38     
#  [82] ROCR_1.0-11          purrr_0.3.4          tensor_1.5          
#  [85] labeling_0.3         patchwork_1.0.1      htmlwidgets_1.6.1   
#  [88] tidyselect_1.1.1     RcppAnnoy_0.0.18     plyr_1.8.5          
#  [91] R6_2.4.1             generics_0.0.2       DBI_1.1.0           
#  [94] mgcv_1.8-31          pillar_1.6.2         whisker_0.4         
#  [97] withr_2.5.0          fitdistrplus_1.1-1   mixsqp_0.3-48       
# [100] survival_3.1-8       abind_1.4-5          tibble_3.1.3        
# [103] future.apply_1.6.0   crayon_1.4.1         KernSmooth_2.23-16  
# [106] utf8_1.1.4           plotly_4.10.1        rmarkdown_2.21      
# [109] grid_3.6.2           data.table_1.14.2    git2r_0.29.0        
# [112] digest_0.6.23        xtable_1.8-4         tidyr_1.1.3         
# [115] httpuv_1.5.2         munsell_0.5.0        viridisLite_0.3.0   
# [118] bslib_0.3.1</code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
