<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Yusha Liu" />


<title>Dissecting tumor transcriptional heterogeneity from single-cell RNA-seq data by generalized binary covariance decomposition (GBCD)</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">gbcd workflow</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephenslab/gbcd-workflow">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Dissecting tumor transcriptional
heterogeneity from single-cell RNA-seq data by generalized binary
covariance decomposition (GBCD)</h1>
<h4 class="author">Yusha Liu</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2023-08-18
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 2
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>gbcd-workflow/analysis/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.0). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomstephenslabgbcdworkflowtreea78ffef34f4f8acb4d78746d587e840e730b7cfdtargetblanka78ffefa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/stephenslab/gbcd-workflow/tree/a78ffef34f4f8acb4d78746d587e840e730b7cfd" target="_blank">a78ffef</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomstephenslabgbcdworkflowtreea78ffef34f4f8acb4d78746d587e840e730b7cfdtargetblanka78ffefa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/stephenslab/gbcd-workflow/tree/a78ffef34f4f8acb4d78746d587e840e730b7cfd" target="_blank">a78ffef</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Untracked files:
    Untracked:  .DS_Store

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/index.Rmd</code>) and HTML
(<code>docs/index.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/gbcd-workflow/blob/a78ffef34f4f8acb4d78746d587e840e730b7cfd/analysis/index.Rmd" target="_blank">a78ffef</a>
</td>
<td>
YushaLiu
</td>
<td>
2023-08-18
</td>
<td>
Add installation instructions
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/gbcd-workflow/f0df13d6b9e88693a76c4022c3b2b2ebac5b5788/docs/index.html" target="_blank">f0df13d</a>
</td>
<td>
YushaLiu
</td>
<td>
2023-08-18
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/gbcd-workflow/blob/a788fb984350bc6d9db743716a58b8710d0996d5/analysis/index.Rmd" target="_blank">a788fb9</a>
</td>
<td>
YushaLiu
</td>
<td>
2023-08-18
</td>
<td>
Update index.Rmd
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/gbcd-workflow/3e5891c7beb248b67c65fa82eceac7e0b6fbfaf7/docs/index.html" target="_blank">3e5891c</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2023-08-09
</td>
<td>
Small fix to the overview page.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/gbcd-workflow/blob/8d6197d4ee9c196f14610955a4f3c8bad99ddede/analysis/index.Rmd" target="_blank">8d6197d</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2023-08-09
</td>
<td>
workflowr::wflow_publish("index.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/gbcd-workflow/0e2b9b8e311b592b8edaa4b25fde7cbb2b840742/docs/index.html" target="_blank">0e2b9b8</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2023-08-09
</td>
<td>
Built the overview page.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/gbcd-workflow/blob/167e5362541382bc277afd68884a45801559b2d3/analysis/index.Rmd" target="_blank">167e536</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2023-08-09
</td>
<td>
Updated the workflowr config.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/gbcd-workflow/blob/db41a579375396dde804dd596aa451d628e87c13/analysis/index.Rmd" target="_blank">db41a57</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2023-08-09
</td>
<td>
Ran wflow_start().
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>This repository contains code and data resources to accompany our
research paper:</p>
<blockquote>
<p>Yusha Liu, Peter Carbonetto, Jason Willwerscheid, Scott A. Oakes, Kay
F. Macleod, and Matthew Stephens (2023). <em>Dissecting tumor
transcriptional heterogeneity from single-cell RNA-seq data by
generalized binary covariance decomposition.</em> bioRxiv <a
href="https://doi.org/10.1101/2023.08.15.553436">doi:10.1101/2023.08.15.553436</a>.</p>
</blockquote>
<p>We provide the following resources:</p>
<ol style="list-style-type: decimal">
<li><p>A <a href="gbcd_hnscc_intro.html">vignette</a> that shows how to
use GBCD to dissect tumor transcriptional heterogeneity through analysis
of multi-tumor single-cell RNA-seq (scRNA-seq) data. We illustrate this
using a head and neck squamous cell carcinoma (HNSCC) dataset analyzed
in our research paper.</p></li>
<li><p>The <a
href="https://github.com/stephenslab/gbcd-workflow">scripts</a> that
reproduce the results and figures presented in the research
paper.</p></li>
</ol>
</div>
<div id="installation-instructions" class="section level2">
<h2>Installation instructions</h2>
<p>Implementing GBCD requires installing the R packages <a
href="https://github.com/stephens999/ashr">ashr (version 2.2-54)</a>, <a
href="https://stephenslab.github.io/ebnm/">ebnm (version 1.0-42)</a> and
<a href="https://github.com/willwerscheid/flashier">flashier (version
0.2.50)</a>, which were previously developed by <a
href="https://stephenslab.uchicago.edu">our lab</a>. All the analyses in
this research paper were performed in <a
href="https://www.r-project.org">R</a> (version 4.1.0).</p>
</div>
<div id="citing-this-work" class="section level2">
<h2>Citing this work</h2>
<p>If you find any material in this repository useful for your work,
please cite our research paper.</p>
</div>
<div id="license" class="section level2">
<h2>License</h2>
<p>All source code and software in this repository are made available
under the terms of the <a
href="https://opensource.org/licenses/mit-license.html">MIT
license</a>.</p>
</div>
<div id="whats-included-in-the-git-repository" class="section level2">
<h2>What’s included in the git repository</h2>
<p>See <a href="https://github.com/stephenslab/gbcd-workflow">here</a>
for the source repository. This is what you will find in the
repository:</p>
<pre><code>├── analysis
├── code
├── docs
├── hnscc
├── pdac
└── simulations
</code></pre>
<ul>
<li><p>The <strong>analysis</strong> contains R Markdown source files
for the workflowr website, including a <a
href="gbcd_hnscc_intro.html">vignette</a> illustrating how to apply GBCD
to analyze your own scRNA-seq data from <em>malignant cells</em>
collected from multiple patients and/or studies.</p></li>
<li><p>The <strong>code</strong> directory contains the R source code to
implement GBCD, using the functions defined in the <a
href="https://stephenslab.github.io/ebnm/">ebnm</a> and <a
href="https://github.com/willwerscheid/flashier">flashier</a> packages
previously developed by our lab.</p></li>
<li><p>The <strong>docs</strong> directory contains webpages generated
from the R Markdown files in the analysis directory.</p></li>
<li><p>The <strong>hnscc</strong> directory stores the preprocessed
HNSCC dataset analyzed in our research paper, and the results and
analysis scripts of the HNSCC data to reproduce the relevant figures in
the research paper.</p></li>
<li><p>The <strong>pdac</strong> directory stores the results and
analysis scripts of the pancreatic cancer adenocarcinoma data to
reproduce the relevant figures in the research paper.</p></li>
<li><p>The <strong>simulations</strong> directory stores the results and
analysis scripts related to the simulation study to reproduce the
relevant figures in the research paper.</p></li>
</ul>
<p>Please note that running these scripts may give you results that are
slightly different from those presented in the paper (which were
generated much earlier), particularly the GBCD results, due to version
updates of the model fitting algorithm. However, the conclusions
reported in the paper remain unaffected.</p>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
